### ISSUE TITLE
Feature request: Add Dropout

### ISSUE THREAD

**Issue by hypnopump** [FlagOpen/FlagAttention#7]
The pytorch base implementation of [`scaled_dot_product_attention`](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html#torch.nn.functional.scaled_dot_product_attention) provides dropout as an arg. Fusing it into the triton kernel would replicate that functionality, as dropout is applied to the attention scores, not the output. 

In the CUDA version, it is supported (here](https://github.com/Dao-AILab/flash-attention/blob/02ac572f3ffc4f402e4183aaa6824b45859d3ed3/csrc/flash_attn/src/flash_fwd_kernel.h#L1142)
There have been attempts at integrating into triton [before](https://github.com/Dao-AILab/flash-attention/issues/116)

**Comment by iclementine** [comment]
Yes. It is on our plan. 

Since triton provides pseudo-random generator now we can implement an memory efficient flash attention with dropout without having to save the dropout mask(since it requires O(n^2) memory). The essence is to re-generate the same dropout masking as is used in the forward pass.

**Comment by hypnopump** [comment]
I have implemented a prototype which [seems to work here](https://github.com/hypnopump/FlagAttention/tree/add_dropout) but it's hard to test correctness without separately implementing the dropout layer and checking, as uses a different random seed than torch.

**Comment by iclementine** [comment]
Yes, testing for randomness is tricky. There is no proper `all_close` tests for operators with randomness. But I think dropout can be testing in the following ways:

1. re-generation of mask;
2. distribution test of mask.

This method can be applied to a separate dropout operator. It can also be applied to the dropout part of a more-complex-operator, but the overall testing for correctness is more complicated then operators without randomness involved.

**Comment by iclementine** [comment]
finally done in https://github.com/FlagOpen/FlagAttention/pull/23
